File: frontend/src/utils/eligibility.js
/**
 * Client-side eligibility checker mirroring the backend Python engine.
 * Used for instant UI feedback when listing drives.
 */
export function checkEligibility(student, drive) {
  if (!student || !drive) {
    return { eligible: false, reasons: ["Profile or drive information missing."] };
  }

  const reasons = [];

  // 1. Placed status check
  if (student.placed) {
    reasons.append ? reasons.push("You are already placed.") : reasons.push("You have already been placed in a company.");
  }

  // 2. Drive active status
  if (!drive.is_active) {
    reasons.push("Recruitment drive is no longer active.");
  }

  // 3. Deadline check
  const deadlineDate = new Date(drive.application_deadline);
  if (new Date() > deadlineDate) {
    reasons.push(`Application deadline (${deadlineDate.toLocaleString()}) has passed.`);
  }

  // 4. CGPA check
  if (student.cgpa < drive.min_cgpa) {
    reasons.push(`Your CGPA (${student.cgpa.toFixed(2)}) is below the required ${drive.min_cgpa.toFixed(2)}.`);
  }

  // 5. Backlogs check
  if (student.active_backlogs > drive.max_backlogs) {
    reasons.push(`Your active backlogs (${student.active_backlogs}) exceed the maximum allowed (${drive.max_backlogs}).`);
  }

  // 6. Branch check
  const studentBranch = (student.branch || "").toUpperCase().trim();
  const eligibleBranches = (drive.eligible_branches || []).map(b => b.toUpperCase().trim());
  if (eligibleBranches.length > 0 && !eligibleBranches.includes(studentBranch)) {
    reasons.push(`Your branch '${student.branch}' is not eligible. Eligible: ${drive.eligible_branches.join(", ")}.`);
  }

  // 7. Graduation year check
  const studentGradYear = parseInt(student.graduation_year);
  const eligibleGradYears = (drive.eligible_grad_years || []).map(y => parseInt(y));
  if (eligibleGradYears.length > 0 && !eligibleGradYears.includes(studentGradYear)) {
    reasons.push(`Your graduation year (${student.graduation_year}) is not eligible. Eligible: ${drive.eligible_grad_years.join(", ")}.`);
  }

  return {
    eligible: reasons.length === 0,
    reasons,
  };
}

File: backend/app/eligibility.py
from datetime import datetime
from typing import TypedDict, List
from backend.app.models import Student, Drive

class EligibilityResult(TypedDict):
    eligible: bool
    reasons: List[str]

def check_student_eligibility(student: Student, drive: Drive) -> EligibilityResult:
    """
    Centralized eligibility checker.
    Returns whether a student is eligible for a recruitment drive, along with reasons if ineligible.
    """
    reasons = []

    # 1. Drive active status
    if not drive.is_active:
        reasons.append("Recruitment drive is no longer active.")

    # 2. Application deadline check
    if datetime.utcnow() > drive.application_deadline:
        reasons.append(f"Application deadline ({drive.application_deadline.strftime('%Y-%m-%d %H:%M')}) has passed.")

    # 3. CGPA check (0 to 10 scale)
    if student.cgpa < drive.min_cgpa:
        reasons.append(f"Your CGPA ({student.cgpa:.2f}) is below the minimum required ({drive.min_cgpa:.2f}).")

    # 4. Backlog check
    if student.active_backlogs > drive.max_backlogs:
        reasons.append(f"Your active backlogs ({student.active_backlogs}) exceed the maximum allowed ({drive.max_backlogs}).")

    # 5. Branch eligibility (case-insensitive checking)
    eligible_branches_lower = [b.lower().strip() for b in drive.eligible_branches]
    student_branch_lower = student.branch.lower().strip()
    if student_branch_lower not in eligible_branches_lower:
        reasons.append(f"Your branch '{student.branch}' is not eligible for this drive. Eligible branches: {', '.join(drive.eligible_branches)}.")

    # 6. Graduation year check
    if student.graduation_year not in drive.eligible_grad_years:
        reasons.append(f"Your graduation year ({student.graduation_year}) is not eligible. Eligible years: {', '.join(map(str, drive.eligible_grad_years))}.")

    # 7. Placed status check
    if student.placed:
        reasons.append("You have already been placed in a company.")

    return {
        "eligible": len(reasons) == 0,
        "reasons": reasons
    }
